test/e2e-gloo-tests.sh

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env bash
+
+# This script runs e2e tests for Canary initialization, analysis and promotion
+# Prerequisites: Kubernetes Kind, Helm and NGINX ingress controller
+
+set -o errexit
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
+export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
+
+echo '>>> Creating test namespace'
+kubectl create namespace test
+
+echo ">>> Downloading Gloo CLI"
+curl -SsL https://github.com/solo-io/gloo/releases/download/v0.13.25/glooctl-linux-amd64 > glooctl
+chmod +x glooctl
+
+echo '>>> Installing load tester'
+kubectl -n test apply -f ${REPO_ROOT}/artifacts/loadtester/
+kubectl -n test rollout status deployment/flagger-loadtester
+
+echo '>>> Initialising canary'
+kubectl apply -f ${REPO_ROOT}/test/e2e-workload.yaml
+./glooctl add route --path-prefix / --upstream-group-name podinfo --upstream-group-namespace test
+
+cat <<EOF | kubectl apply -f -
+apiVersion: flagger.app/v1alpha3
+kind: Canary
+metadata:
+  name: podinfo
+  namespace: test
+spec:
+  targetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: podinfo
+  progressDeadlineSeconds: 60
+  service:
+    port: 9898
+  canaryAnalysis:
+    interval: 15s
+    threshold: 15
+    maxWeight: 30
+    stepWeight: 10
+    metrics:
+    - name: envoy-success-rate
+      threshold: 99
+      interval: 1m
+      query: |
+        sum(rate(envoy_cluster_upstream_rq{kubernetes_namespace="gloo-system", gloo="gateway-proxy", 
```

Comment:
The query must contain the deployment namespace and name labels and it should be moved in the metrics package. There are two builtin metrics that Goo should implement `request-success-rate` and `request-duration`, please see https://docs.flagger.app/how-it-works#http-metrics and implementation examples here https://github.com/weaveworks/flagger/tree/master/pkg/metrics

Reply:
I've refactored the metrics and added a factory similar to the router one. Please merge the master into this branch to get the latest metrics package.